ntoskrnl/io/iomgr/deviface.c

```diff
@@ -28,190 +29,499 @@ IopGetDeviceObjectFromDeviceInstance(PUNICODE_STRING DeviceInstance);
+ */
+static
+NTSTATUS
+IopSeparateSymbolicLink(
+    _In_ PCUNICODE_STRING SymbolicLinkName,
+    _Out_opt_ PUNICODE_STRING PrefixString,
+    _Out_opt_ PUNICODE_STRING MungedString,
+    _Out_opt_ PUNICODE_STRING GuidString,
+    _Out_opt_ PUNICODE_STRING ReferenceString,
+    _Out_opt_ PBOOLEAN ReferenceStringPresent,
+    _Out_opt_ LPGUID InterfaceClassGuid)
+{
+    const UNICODE_STRING KernelModePrefix = RTL_CONSTANT_STRING(L"\\??\\");
+    const UNICODE_STRING UserModePrefix = RTL_CONSTANT_STRING(L"\\\\?\\");
+    const UNICODE_STRING MungedPrefix = RTL_CONSTANT_STRING(L"##?#");
+    UNICODE_STRING MungedStringReal, GuidStringReal, ReferenceStringReal;
+    UNICODE_STRING DeviceString;
+    UNICODE_STRING LinkNameNoPrefix;
+    USHORT i, ReferenceStringOffset;
+    NTSTATUS Status = STATUS_SUCCESS;
 
-    RtlInitUnicodeString(&BaseKeyU, BaseKeyString);
+    DPRINT("Symbolic link is '%wZ'\n", SymbolicLinkName);
 
-    /* Open the DeviceClasses key */
-    InitializeObjectAttributes(&ObjectAttributes,
-                               &BaseKeyU,
-                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
-                               NULL,
-                               NULL);
-    Status = ZwOpenKey(&ClassesKey,
-                       DesiredAccess | KEY_ENUMERATE_SUB_KEYS,
-                       &ObjectAttributes);
-    if (!NT_SUCCESS(Status))
+    /* Symbolic link name looks like \??\ACPI#PNP0501#1#{GUID}\ReferenceString */
+    /* Make sure it starts with the expected prefix */
+    if (!RtlPrefixUnicodeString(&KernelModePrefix, SymbolicLinkName, FALSE) &&
+        !RtlPrefixUnicodeString(&UserModePrefix, SymbolicLinkName, FALSE))
     {
-        DPRINT1("Failed to open %wZ\n", &BaseKeyU);
-        goto cleanup;
+        DPRINT1("Invalid link name '%wZ'\n", SymbolicLinkName);
+        return STATUS_INVALID_PARAMETER;
     }
 
-    StartPosition = wcschr(SymbolicLinkName->Buffer, L'{');
-    EndPosition = wcschr(SymbolicLinkName->Buffer, L'}');
-    if (!StartPosition || !EndPosition || StartPosition > EndPosition)
+    /* Sanity checks */
+    ASSERT(KernelModePrefix.Length == UserModePrefix.Length);
+    ASSERT(SymbolicLinkName->Length >= KernelModePrefix.Length);
+
+    /* Make a version without the prefix for further processing */
+    LinkNameNoPrefix.Buffer = SymbolicLinkName->Buffer + KernelModePrefix.Length / sizeof(WCHAR);
+    LinkNameNoPrefix.Length = SymbolicLinkName->Length - KernelModePrefix.Length;
+    LinkNameNoPrefix.MaximumLength = LinkNameNoPrefix.Length;
+
+    DPRINT("Symbolic link without prefix is '%wZ'\n", &LinkNameNoPrefix);
+
+    /* Find the reference string, if any */
+    for (i = 0; i < LinkNameNoPrefix.Length / sizeof(WCHAR); i++)
     {
-        DPRINT1("Bad symbolic link: %wZ\n", SymbolicLinkName);
-        return STATUS_INVALID_PARAMETER_1;
+        if (LinkNameNoPrefix.Buffer[i] == L'\\')
+        {
+            break;
+        }
```

Comment:
nit:
```suggestion
            break;
```
For the record: I do intend to give my approval for this PR, but only after testbot-results will have been attached here for WHS and for KVM+VBox, which ofc requires our infra to be repaired first.

Reply:
> How much is missing after this PR to make the files replacement actually output sound?

@JoachimHenze I'm not sure how much exactly yet, but for the full replacement, there is incompatibility in our umpnpmgr and most likely setupapi, according to my findings. With 2k3sp2 umpnpmgr.dll (and ServiceMain registry value changed accordingly) the drivers are not unloaded anymore at least, but the sound still does not work. For some reason, MS wdmaud.sys seems to return 0 devices for some reason (because GetNumDevs fails somewhere), but again, I'm not sure exactly it occurs because of our setupapi/umpnpmgr. This needs much more investigations.
Our winmm.dll also might be a reason, as it interacts between wdmaud.drv and other user mode components like dsound.dll.

Reply:
Done.